HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h

```diff
@@ -0,0 +1,100 @@
+#ifndef HeterogeneousCore_CUDAUtilities_interface_device_unique_ptr_h
+#define HeterogeneousCore_CUDAUtilities_interface_device_unique_ptr_h
+
+#include <memory>
+#include <functional>
+
+#include "FWCore/Utilities/interface/Likely.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/allocate_device.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/currentDevice.h"
+
+namespace cudautils {
+  namespace device {
+    namespace impl {
+      // Additional layer of types to distinguish from host::unique_ptr
+      class DeviceDeleter {
+      public:
+        DeviceDeleter() = default;  // for edm::Wrapper
+        DeviceDeleter(int device) : device_{device} {}
+
+        void operator()(void *ptr) {
+          if (LIKELY(device_ >= 0)) {
+            free_device(device_, ptr);
+          }
+        }
+
+      private:
+        int device_ = -1;
+      };
+    }  // namespace impl
+
+    template <typename T>
+    using unique_ptr = std::unique_ptr<T, impl::DeviceDeleter>;
+
+    namespace impl {
+      template <typename T>
+      struct make_device_unique_selector {
+        using non_array = cudautils::device::unique_ptr<T>;
+      };
+      template <typename T>
+      struct make_device_unique_selector<T[]> {
+        using unbounded_array = cudautils::device::unique_ptr<T[]>;
+      };
+      template <typename T, size_t N>
```

Comment:
To disallow making arrays of known bounds similarly to `std::make_unique()`.

Reply:
By the way, it does puzzle me why `std::make_unique<int[5]>()` is not allowed...

It looks like it is because you can do
```c++
std::unique_ptr<int> p(new int);
```
and
```c++
std::unique_ptr<int[]> p(new int[42]);
```
but not
```c++
std::unique_ptr<int[42]> p(new int[42]);
```
and `std::make_unique()` follows that, rather than making `std::make_unique<int[42]>(...)` return a `std::unique_ptr<int[]>`.

So, *if* we have a specialisation for bounded arrays, and we think they could be useful, IMHO it would be OK to have implement `make_device_unique` and `make_host_unique` for bounded arrays.

Reply:
I would follow strictly what `std` does, fearing that anything else would lead to confusion. As of why `std::make_unique<int[5]>()` is disallowed, I'll quote the proposal via stackoverflow
>`T[N]`
>
> As of N3485, `unique_ptr` doesn't provide a partial specialization for `T[N]`. However, users will be strongly tempted to write `make_unique<T[N]>()`. This is a no-win scenario. Returning `unique_ptr<T[N]>` would select the primary template for single objects, which is bizarre. Returning `unique_ptr<T[]>` would be an exception to the otherwise ironclad rule that `make_unique<something>()` returns `unique_ptr<something>`. Therefore, this proposal makes `T[N]` ill-formed here, allowing implementations to emit helpful `static_assert` messages.

https://stackoverflow.com/a/16597578

Reply:
As I wrote, that simply states that `make_unique<int[5]>()` is not supported because `unique_ptr<int[5]>` is not properly specialised.

For a smart pointer that implements a specialisation for arrays with known bound, it would make sense to implement the corresponding `make_` free function.